java/java-psi-api/src/com/intellij/psi/util/OrderedBinaryExpression.java

```diff
@@ -0,0 +1,122 @@
+import com.intellij.psi.tree.IElementType;
+
+import java.util.Arrays;
+
+import static com.intellij.util.ObjectUtils.tryCast;
+
+/**
+ * Ordered binary expression for commutative operations.
+ *
+ * @param <F> First operand
+ * @param <S> Second operand
+ */
+public class OrderedBinaryExpression<F extends PsiExpression, S extends PsiExpression> {
+  private F firstOperand;
+  private IElementType tokenType;
+  private S secondOperand;
+
+  private OrderedBinaryExpression(F firstOperand, IElementType tokenType, S secondOperand) {
+    this.firstOperand = firstOperand;
+    this.tokenType = tokenType;
+    this.secondOperand = secondOperand;
+  }
+
+  public F getFirstOperand() {
+    return firstOperand;
+  }
+
+  public IElementType getTokenType() {
+    return tokenType;
+  }
+
+  public S getSecondOperand() {
+    return secondOperand;
+  }
+
+  /**
+   * Return the items of a binary expression in the order it is specified. It reverses the operator if needed.
+   *
+   * @param <O>          the required expression type
+   * @param node         the supposed binary expression
+   * @param firstClass   the class representing the required expression type
+   * @return the items of a binary expression in the order it is specified. It reverses the operator if needed.
+   */
+  public static <O extends PsiExpression> OrderedBinaryExpression<O, PsiExpression> from(final PsiExpression node, final Class<O> firstClass) {
+    PsiBinaryExpression expression = as(node, PsiBinaryExpression.class);
+    if (expression == null || firstClass == null) return null;
+    O leftOperand = as(expression.getLOperand(), firstClass);
+    O rightOperand = as(expression.getROperand(), firstClass);
+    if (leftOperand != null && rightOperand == null) {
+      return new OrderedBinaryExpression<>(leftOperand, expression.getOperationTokenType(), PsiUtil.skipParenthesizedExprDown(expression.getROperand()));
+    }
+    if (leftOperand == null && rightOperand != null) {
+      IElementType mirroredOperator = mirrorOperator(expression);
+      return new OrderedBinaryExpression<>(rightOperand, mirroredOperator, PsiUtil.skipParenthesizedExprDown(expression.getLOperand()));
+    }
+    return null;
+  }
+
+  /**
+   * Return the items of a binary expression in the order it is specified. It reverses the operator if needed.
+   *
+   * @param <F>          the required expression type
+   * @param <S>          the required expression type
+   * @param node         the supposed binary expression
+   * @param firstClass   the class representing the required expression type
+   * @param secondClass  the class representing the required expression type
+   * @return the items of a binary expression in the order it is specified. It reverses the operator if needed.
+   */
+  public static <F extends PsiExpression, S extends PsiExpression> OrderedBinaryExpression<F, S> from(final PsiExpression node, final Class<F> firstClass, final Class<S> secondClass) {
```

Comment:
Please add explicit Nullable/NotNull annotations to every new public method

Reply:
Done